typescript/hardhat/src/TestAbacusDeploy.ts

```diff
@@ -144,62 +143,74 @@ export class TestAbacusDeploy extends TestDeploy<
-    const checkpoints = await outbox.queryFilter(checkpointFilter);
-    if (!(checkpoints.length === 0 || checkpoints.length === 1))
-      throw new Error("found multiple checkpoints");
-    const fromBlock = checkpoints.length === 0 ? 0 : checkpoints[0].blockNumber;
+  async processOutboundMessages(
+    origin: types.Domain
+  ): Promise<Map<types.Domain, ethers.providers.TransactionResponse[]>> {
+    const responses: Map<types.Domain, ethers.providers.TransactionResponse[]> =
+      new Map();
+    const outbox = this.outbox(origin);
+    const [, checkpointedIndex] = await outbox.latestCheckpoint();
+    const latestIndex = await outbox.count();
+    if (latestIndex.eq(checkpointedIndex)) return responses;
 
     await outbox.checkpoint();
     const [root, index] = await outbox.latestCheckpoint();
-    // If there have been no checkpoints since the last checkpoint, return.
-    if (
-      index.eq(0) ||
-      (checkpoints.length == 1 && index.eq(checkpoints[0].args.index))
-    ) {
-      return;
-    }
-    // Update the Outbox and Inboxes to the latest roots.
+
+    // Update the Inboxes to the latest root.
     // This is technically not necessary given that we are not proving against
     // a root in the TestInbox.
     const validator = await Validator.fromSigner(
-      this.config.signer[domain],
-      domain
+      this.config.signer[origin],
+      origin
     );
     const { signature } = await validator.signCheckpoint(
       root,
       index.toNumber()
     );
 
-    for (const remote of this.remotes(domain)) {
-      const inbox = this.inbox(remote, domain);
+    for (const destination of this.remotes(origin)) {
+      const inbox = this.inbox(origin, destination);
       await inbox.checkpoint(root, index, signature);
     }
 
     // Find all messages dispatched on the outbox since the previous checkpoint.
     const dispatchFilter = outbox.filters.Dispatch();
-    const dispatches = await outbox.queryFilter(dispatchFilter, fromBlock);
+    const dispatches = await outbox.queryFilter(dispatchFilter);
     for (const dispatch of dispatches) {
       const destination = dispatch.args.destinationAndNonce.shr(32).toNumber();
-      if (destination !== domain) {
-        const inbox = this.inbox(destination, domain);
+      if (destination === origin)
+        throw new Error("Dispatched message to local domain");
+      const inbox = this.inbox(origin, destination);
+      const status = await inbox.messages(dispatch.args.messageHash);
+      if (status !== types.MessageStatus.PROCESSED) {
         await inbox.setMessageProven(dispatch.args.message);
-        await inbox.testProcess(dispatch.args.message);
+        const response = await inbox.testProcess(dispatch.args.message);
+        if (!responses.get(destination)) {
+          responses.set(destination, [response]);
+        } else {
+          responses.get(destination)!.push(response);
+        }
```

Comment:
```suggestion
        const destinationResponses = responses.get(destination) || [];
        responses.set(destination, destinationResponses.push(response));
```

fun fact I think we will have `emplace` soon?
https://github.com/tc39/proposal-upsert